index.rst

```diff
@@ -48,9 +48,118 @@
+
+Currently, we use the ``provisionator`` (``769:769``) account to perform
+those actions, which are privileged.  In order to do that, the container
+build writes sudoers rules to allow the provisionator to act.  It also
+allows the provisionator to write a rule (via a script) that allows the
+provisionator to become the target user before starting Jupyterlab.
+
+This requires ``sudo`` in order to work, and therefore the Lab
+containers cannot be run (at least in environments that require
+filesystem provisioning) with ``allowPrivilegeEscalation: False``.  This
+violates the third item of section 4.1 of `SQR-048`_.
+
+.. _SQR-048: https://sqr-048.lsst.io/
+
+The Partial Solution
+====================
+
+A year or more ago we introduced a ``NO_SUDO`` flag environment variable
+that can be passed into the container, which bypasses all of the sudo
+steps to escalate privilege.
+
+This comes with one major caveat: it only works if there is a
+pre-provisioned user home directory.  This is fine at NCSA (which is
+where the rule was intended for use) since user home space is
+provisioned as part of account creation and NCSA forces all RSP users to
+have NCSA accounts; it is not fine in a GKE environment using GitHub as
+its authentication back-end.  We have recently proven that everything
+still works as expected (assuming a pre-provisioned filesystem) if the
+container is run with ``allowPrivilegeEscalation: False`` in its
+Kubernetes ``securityContext``.
+
+Recently, in order to allow loading the Multus CNI into containers in
+Rubin-managed Kubernetes environments, necessary for control of
+instruments on multicast networks with SAL, Josh Hoblitt introduced a
+container that must be run as a privileged initContainer inside the
+user's Lab pod before starting the Lab.  This change is already present
+in the mainline Nublado environment.
+
+A Fuller Solution
+=================
+
+Since we've worked out how to run privileged initContainers, all we
+would need to do is to move the provisioning steps into a second
+initContainer, and invoke it (or not) at user pod creation time.  We
+could either restrict its privilege still further by doing the same set
+of semi-privileged user plus sudoers rules we currently do, or just let
+the initContainer run as root, which would simplify the logic within
+considerably.
+
+Then, in all cases, we could invoke the Lab pod directly with the
+correct userid and set of groups, having provisioned those in ConfigMaps
+to mount into ``/etc/passwd`` and ``/etc/group`` (``NO_SUDO`` already
+does this), with the only piece that still requires privilege being the
+creation of user filesystems.
+
+This will be an initContainer that mounts our NFS share volume with
+``no_root_squash`` and effectively runs as root while doing so in order
+to create the user spaces.  That's a little scary, but no scarier than
+what we are doing today, particularly if we restrict the provisionator's
+actions with sudoers rules.
```

Comment:
Is whether or not to use `root_squash` entirely under the control of the client and thus under the control of the pod volumes configuration?

I don't think sudo buys us any security here versus just running the container as root.

Reply:
I think you're basically right.  I would have no objection to running the homedir provisioner as root, if that's all it did.